src/plugins/intel_cpu/src/nodes/ngram.cpp

```diff
@@ -0,0 +1,164 @@
+        if (!ngram) {
+            IE_THROW(Unexpected) << "Wrong operation type";
+        }
+        return std::make_shared<NgramShapeInfer>(ngram->get_k());
+    }
+private:
+    std::shared_ptr<ov::Node> m_op;
+};
+}   // namespace
+
+bool Ngram::isSupportedOperation(const std::shared_ptr<const ov::Node>& op, std::string& errorMessage) noexcept {
+    try {
+        const auto ngram = ov::as_type_ptr<const NgramNode>(op);
+        if (!ngram) {
+            errorMessage = "Only Ngram from CPU internal opset is supported";
+            return false;
+        }
+    } catch (...) {
+        return false;
+    }
+
+    return true;
+}
+
+Ngram::Ngram(const std::shared_ptr<ov::Node>& op, const GraphContext::CPtr& context)
+    : Node(op, context, NgramShapeInferFactory(op)) {
+    std::string errorMessage;
+    if (!isSupportedOperation(op, errorMessage)) {
+        IE_THROW(NotImplemented) << errorMessage;
+    }
+
+    const auto ngram = ov::as_type_ptr<const NgramNode>(op);
+    k = ngram->get_k();
+    leftPad = k % 2 == 0 ? (k - 1) / 2 : k / 2;
+    rightPad = k / 2;
+
+    const auto& windowStrideDim = ngram->get_input_partial_shape(0)[1];
+    if (windowStrideDim.is_static()) {
+        windowStride = windowStrideDim.get_length();
+        windowSize = k * windowStride;
+        leftPaddingSize = windowStride * leftPad;
+        rightPaddingSize = windowStride * rightPad;
+    }
+}
+
+void Ngram::initSupportedPrimitiveDescriptors() {
+    if (!supportedPrimitiveDescriptors.empty())
+        return;
+
+    addSupportedPrimDesc({{LayoutType::ncsp, InferenceEngine::Precision::FP32},
+                          {LayoutType::ncsp, InferenceEngine::Precision::I32}},
+                         {{LayoutType::ncsp, InferenceEngine::Precision::FP32}},
+                         ref_any,
+                         isDynamicNode());
+}
+
+void Ngram::prepareParams() {
+    const auto& srcDataDims = getParentEdgeAt(0)->getMemoryPtr()->getStaticDims();
+    const auto& srcIndicesDims = getParentEdgeAt(1)->getMemoryPtr()->getStaticDims();
+
+    idcesShapeSize = std::accumulate(srcIndicesDims.begin(), srcIndicesDims.end(), 1, std::multiplies<size_t>());
+    idcesStride = getParentEdgeAt(1)->getMemoryPtr()->GetDescWithType<BlockedMemoryDesc>()->getStrides()[0];
+    numIdces = srcIndicesDims[0];
+
+    windowStride = srcDataDims[1];
+    windowSize = k * windowStride;
+    leftPaddingSize = windowStride * leftPad;
+    rightPaddingSize = windowStride * rightPad;
+}
+
+void Ngram::execute(dnnl::stream strm) {
+    auto* srcData = reinterpret_cast<const float*>(getParentEdgeAt(0)->getMemoryPtr()->GetPtr());
+    auto* srcIndices = reinterpret_cast<const int*>(getParentEdgeAt(1)->getMemoryPtr()->GetPtr());
+    auto* dstData = reinterpret_cast<float*>(getChildEdgeAt(0)->getMemoryPtr()->GetPtr());
+
+    std::vector<size_t> batchLenghts{0};
+    batchLenghts.reserve(numIdces + 1);
+    for (size_t i = idcesStride; i < idcesShapeSize; i += idcesStride) {
```

Comment:
Looks like that it's easy to support INT64 here. Just need to dub this loop. Could you please do so?

Reply:
In terms of semantic, I64 indices makes sense only in the case when number of embeddings more than I32::max. I don't think that we can meet such cases in real scenario. Given the above, do we really need I64 support here?
If yes, could you please clarify what means "Just need to dub this loop"? Didn't get it.

Reply:
We are going to add optional support of INT64 and to avoid excessive Convert i32->i64 better to add native support in node.
The possible solution:
`
auto mem = getParentEdgeAt(1)->getMemoryPtr();
if (mem->getDesc().getPrecision() == Precision::I64) {
    auto* srcIndices = reinterpret_cast<const int64_t*>(mem->GetPtr());
    for (size_t i = idcesStride; i < idcesShapeSize; i += idcesStride) {
        if (srcIndices[i - idcesStride] != srcIndices[i]) {
            batchLenghts.push_back(i / idcesStride);
        }
    }
} else if (mem->getDesc().getPrecision() == Precision::I32) {
    auto* srcIndices = reinterpret_cast<const int32_t*>(mem->GetPtr());
    for (size_t i = idcesStride; i < idcesShapeSize; i += idcesStride) {
        if (srcIndices[i - idcesStride] != srcIndices[i]) {
            batchLenghts.push_back(i / idcesStride);
        }
    }
}`
or use conditional compilation switch.

Reply:
Thank you for the detailed explanation! Will follow your suggestion